### PR TITLE
fix: correct handling of polarizations from UVBeam

### PR DESCRIPTION
**Pull request by steven-murray**
This PR updates the behaviour of vis_cpu to be more consistent.
In particular, when ``polarized=False``, we explicitly take the sqrt
of the power beam for a particular feed. We raise errors for incompatible
inputs.

BREAKING CHANGE: previous inputs for the beam will now sometimes
raise errors, as they were inconsistent. In particular, input
``beam_list`` to ``vis_cpu`` must be efield type if using `polarized=True`
and `power` type if not.

### CODE REVIEW

**Comment by steven-murray** [comment]
@philbull there are two remaining errors in the tests. One is for `test_beam_interpolation`, and it is fixed if I switch out the `EllipticalBeam` for a pure gaussian `AnalyticBeam`, so I think it has more to do with the implementation of `EllipticalBeam` than anything else.

The other is the comparison of vis_cpu to pyuvsim, which now is not matching well at all. I can give it a closer look, but I thought you might spot the error more quickly, since you wrote the test.

**Comment by r-pascua** [comment]
I figured out the problem with the comparison test&mdash;it has to do with how the beam is constructed. Since we're interpolating the `UVBeam` object directly, the unpolarized (power beam) case needs to take the square root of the interpolated beam (this is done in the function that grids the beam, but it isn't done in the actual call to `vis_cpu`). This change just needs to be made on line 253 in `vis_cpu.py`.

I'll see if I can figure out the other problem...

File: src/vis_cpu/conversions.py [hunk 1]
```diff
@@ -258,7 +267,57 @@ def equatorial_to_eci_coords(ra, dec, obstime, location, unit="rad", frame="icrs
...
+        if uvbeam.beam_type != "efield":
+            raise ValueError("Beam type must be efield")
+        uvbeam_ = uvbeam
+    elif uvbeam.beam_type == "efield":
+        uvbeam_ = uvbeam.copy() if isinstance(uvbeam, UVBeam) else deepcopy(uvbeam)
+        # Analytic beams have no concept of feeds, so assume they have a "single" feed
+        if getattr(uvbeam_, "Nfeeds", 1) > 1:
+            uvbeam_.select(feeds=[use_feed])
+
+        uvbeam_.efield_to_power()
```
Comment (r-pascua) [inline src/vis_cpu/conversions.py:296]: I think we want to call this method with `calc_cross_pols=False` as an argument. I think this should fix the failing test. We're not allowing the use of cross-polarized power beams, so this fix should be OK. That said, this seems a little bit like a corner case that's not caught in the `UVBeam` code&mdash;it looks like the `efield_to_power` method doesn't account for the case where the user asks for power beams when there's only one feed and `calc_cross_pols` isn't set to `False` (in the sense that we should expect the resulting beam to be real-valued, but it doesn't enter a clause that explicitly sets it to be real-valued).

File: src/vis_cpu/conversions.py [hunk 2]
```diff
@@ -258,7 +267,57 @@ def equatorial_to_eci_coords(ra, dec, obstime, location, unit="rad", frame="icrs
...
+
+    elif getattr(uvbeam, "Npols", 1) > 1:
+        pol = uvutils.polstr2num(use_pol)
+
+        if pol not in uvbeam.polarization_array:
+            raise ValueError(
+                f"You want to use {use_feed} feed, but it does not exist in the UVBeam"
+            )
```
Comment (r-pascua) [inline src/vis_cpu/conversions.py:304]: I don't think we need to include this check? It looks like this is something that's already covered in the `select` method.

Comment (steven-murray) [reply src/vis_cpu/conversions.py:304]: I'd still prefer to keep this here to make it more obvious for the user.

File: src/vis_cpu/conversions.py (hunk 1 above)
Comment (bhazelton) [reply src/vis_cpu/conversions.py]: If there's only one feed there is no such thing as a cross pol. One feed yields one power beam no matter what.

Comment (bhazelton) [reply src/vis_cpu/conversions.py]: Oh! I just read that more carefully. Yes, I should add some logic to make it real valued if there's only one feed.

File: src/vis_cpu/vis_cpu.py [hunk 3]
```diff
@@ -108,21 +108,17 @@ def vis_cpu(
     beam_list : list of UVBeam, optional
         If specified, evaluate primary beam values directly using UVBeam
         objects instead of using pixelized beam maps (``bm_cube`` will be
-        ignored if ``beam_list`` is not ``None``).
+        ignored if ``beam_list`` is not ``None``). Note that if `polarized` is True,
+        these beams must be efield beams, and conversely they must be power beams
```
Comment (bhazelton) [inline src/vis_cpu/vis_cpu.py:112]: Need another word or two between "conversely" and "they must". Maybe "conversely if `polarized` is False they must..."

File: tests/test_beams.py [hunk 4]
```diff
@@ -188,3 +287,95 @@ def test_beam_interpolation_pol():
...
+            freq[0],
+            eq2tops,
+            crd_eq,
+            sky_flux[:, 0],
+            beam_list=beam_list_unpol,
+            precision=2,
+            polarized=True,
+        )
+
+        assert "beam type must be efield" in str(e)
```
Comment (bhazelton) [inline tests/test_beams.py:306]: you can make this shorter if you use the "match" parameter in `pytest.raises` like this:
```suggestion
    with pytest.raises(ValueError, match="beam type must be efield"):
        vis_cpu(
            antpos,
            freq[0],
            eq2tops,
            crd_eq,
            sky_flux[:, 0],
            beam_list=beam_list_unpol,
            precision=2,
            polarized=True,
        )
```
And similarly in other examples below.

File: src/vis_cpu/conversions.py [hunk 5]
```diff
@@ -258,7 +267,60 @@ def equatorial_to_eci_coords(ra, dec, obstime, location, unit="rad", frame="icrs
     return pra, pdec
 
 
-def uvbeam_to_lm(uvbeam, freqs, n_pix_lm=63, polarized=False, **kwargs):
+def prepare_beam(
+    uvbeam: UVBeam, polarized: bool = False, use_feed: Literal["x", "y"] = "x"
```
Comment (bhazelton) [inline src/vis_cpu/conversions.py:271]: do you want to allow for "n" or "e" if `x_orientation` is set on the UVBeam object?

Comment (steven-murray) [reply src/vis_cpu/conversions.py:271]: Not right now.

File: src/vis_cpu/vis_cpu.py (hunk 3 above)
Comment (steven-murray) [reply src/vis_cpu/vis_cpu.py:112]: Thanks, done.

File: tests/test_beams.py (hunk 4 above)
Comment (steven-murray) [reply tests/test_beams.py:306]: Ah, didn't know this! Nice :-)
